proxy/http2/HTTP2.cc

```diff
@@ -484,246 +498,162 @@ convert_from_2_to_1_1_header(HTTPHdr *headers)
-      if (p == out) {
-        // no progress was made, header was too big for the buffer, skipping for now
-        continue;
-      }
-      if (!cont) {
-        // Parsing a part of headers is done
-        cont = true;
-        return p - out;
-      } else {
-        // Parse error
-        return -1;
-      }
-    }
-    p += len;
-  }
-
-  // Parsing all headers is done
-  return p - out;
+  // TODO: It would be better to split Cookie header value
+  return hpack_encode_header_block(handle, out, out_len, in);
 }
 
 /*
  * Decode Header Blocks to Header List.
  */
 int64_t
-http2_decode_header_blocks(HTTPHdr *hdr, const uint8_t *buf_start, const uint8_t *buf_end, Http2IndexingTable &indexing_table,
+http2_decode_header_blocks(HTTPHdr *hdr, const uint8_t *buf_start, const uint32_t buf_len, HpackHandle &handle,
                            bool &trailing_header)
 {
-  const uint8_t *cursor = buf_start;
-  HdrHeap *heap = hdr->m_heap;
-  HTTPHdrImpl *hh = hdr->m_http;
-  bool header_field_started = false;
+  const MIMEField *field;
+  const char *value;
+  int len;
   bool is_trailing_header = trailing_header;
+  int64_t result = hpack_decode_header_block(handle, hdr, buf_start, buf_len);
 
-  while (cursor < buf_end) {
-    int64_t read_bytes = 0;
+  if (result < 0) {
+    return result;
+  }
 
-    // decode a header field encoded by HPACK
-    MIMEField *field = mime_field_create(heap, hh->m_fields_impl);
-    MIMEFieldWrapper header(field, heap, hh->m_fields_impl);
-    HpackFieldType ftype = hpack_parse_field_type(*cursor);
 
-    switch (ftype) {
-    case HPACK_FIELD_INDEX:
-      read_bytes = decode_indexed_header_field(header, cursor, buf_end, indexing_table);
-      if (read_bytes == HPACK_ERROR_COMPRESSION_ERROR) {
-        return HPACK_ERROR_COMPRESSION_ERROR;
-      }
-      cursor += read_bytes;
-      header_field_started = true;
-      break;
-    case HPACK_FIELD_INDEXED_LITERAL:
-    case HPACK_FIELD_NOINDEX_LITERAL:
-    case HPACK_FIELD_NEVERINDEX_LITERAL:
-      read_bytes = decode_literal_header_field(header, cursor, buf_end, indexing_table);
-      if (read_bytes == HPACK_ERROR_COMPRESSION_ERROR) {
-        return HPACK_ERROR_COMPRESSION_ERROR;
-      }
-      cursor += read_bytes;
-      header_field_started = true;
-      break;
-    case HPACK_FIELD_TABLESIZE_UPDATE:
-      if (header_field_started) {
-        return HPACK_ERROR_COMPRESSION_ERROR;
+  MIMEFieldIter iter;
+  unsigned int expected_pseudo_header_count = 4;
+  unsigned int pseudo_header_count = 0;
+
+  if (is_trailing_header) {
+    expected_pseudo_header_count = 0;
+  }
+  for (field = hdr->iter_get_first(&iter); field != NULL; field = hdr->iter_get_next(&iter)) {
+    value = field->name_get(&len);
+    // Pseudo headers must appear before regular headers
+    if (len && value[0] == ':') {
+      ++pseudo_header_count;
+      if (pseudo_header_count > expected_pseudo_header_count) {
+        return -result;
```

Comment:
Is it necessary to return this? Those values are used in `decode_header_blocks` in Http2Stream.h to detect COMPRESSION_ERROR or PROTOCOL_ERROR.
How about returning Http2ErrorCode directly?